play-json/shared/src/main/scala/JsResult.scala

```diff
@@ -40,27 +49,46 @@ object JsError {
   }
 
   def toJson(e: JsError): JsObject = toJson(e.errors, false)
+
   def toJson(errors: Seq[(JsPath, Seq[JsonValidationError])]): JsObject = toJson(errors, false)
+
   //def toJsonErrorsOnly: JsValue = original // TODO
   def toFlatForm(e: JsError): Seq[(String, Seq[JsonValidationError])] = e.errors.map { case (path, seq) => path.toJsonString -> seq }
 
   private def toJson(errors: Seq[(JsPath, Seq[JsonValidationError])], flat: Boolean): JsObject = {
-    val argsWrite = Writes.traversableWrites[Any](Writes.anyWrites)
-    errors.foldLeft(Json.obj()) { (obj, error) =>
-      obj ++ Json.obj(error._1.toJsonString -> error._2.foldLeft(Json.arr()) { (arr, err) =>
-        arr :+ Json.obj(
-          "msg" -> (if (flat) err.message else Json.toJson(err.messages)),
-          "args" -> Json.toJson(err.args)(argsWrite)
-        )
-      })
+    errors.foldLeft(JsObject.empty) { (obj, error) =>
+      obj ++ JsObject(Seq(error._1.toJsonString -> error._2.foldLeft(JsArray.empty) { (arr, err) =>
+        val msg = JsArray({
+          if (flat) Seq(JsString(err.message))
+          else err.messages.map(JsString(_))
+        })
+        arr :+ JsObject(Seq(
+          "msg" -> msg,
+          "args" -> JsArray(err.args.map(toJson))
+        ))
+      }))
     }
   }
+
+  /**
+   * Serializer for Any, used for the args of errors.
+   */
+  private def toJson(a: Any): JsValue = a match {
```

Comment:
Directly use this internal conversion, rather than doing it through an internal `Writes` instance (never expose in the API)